### PR TITLE
[ML] Make GetJobStats work with arbitrary wildcards and groups

### LINKED ISSUE #34745: [ML] GetJobStatsAction does not match open jobs correctly
The endpoint accepts a comma separated list of job Ids, group Ids, wildcarded Ids or `_all` which is then [expanded](https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportGetJobsStatsAction.java#L72) into individual job Ids. But the [task match function](https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/GetJobsStatsAction.java#L97) works on the unexpanded job id in the request rather than each of the expanded Ids. `_all` is handled correctly.

There are test cases covering getting the job stats for multiple jobs but the tests pass because the code assumes the job is closed then goes and gets the job stats for the job as if it was closed.

### PR DESCRIPTION
**Pull request by henriquegoncalves98**
Fixing match function to use expandedJobId instead of a single jobid.

Closes #34745

### CODE REVIEW

**Comment by up201608320** [comment]
In my commit instead of 'expandedJobsIds[i]' it's expandedJobsIds.get(i). All tests pass with the code submited.

**Review by benwtrent — COMMENTED** [review]
I am not sure what which tests @davidkyle was mentioning in the issue. 

I know we have tests here: https://github.com/elastic/elasticsearch/blob/master/client/rest-high-level/src/test/java/org/elasticsearch/client/MachineLearningIT.java but those are more for ensuring the client acts correctly, not that the job stats are handled appropriately.

**Comment by benwtrent** [comment]
Jenkins test this please

**Comment by benwtrent** [comment]
@up201608320 Thanks for taking a look at this issue. Looking at `elasticsearch-ci-2` it seems that there is a yml test that is failing. I found the test failure by selecting the details link, looking at the full console output, and then searching the page for `REPRODUCE` (case sensitive). That turned up a yml test that is failing.

You can rerun these with:
```
./gradlew :x-pack:plugin:integTestRunner -Dtests.seed=E912E5024622B77D -Dtests.class=org.elasticsearch.xpack.test.rest.XPackRestIT -Dtests.method="test {p0=ml/job_groups/Test close job API}"
```
Relevant test is here: 
https://github.com/elastic/elasticsearch/blob/4103d3b9ec5186aa74c987f72d65c99ee27f4c2b/x-pack/plugin/src/test/resources/rest-api-spec/test/ml/job_groups.yml#L242

I ran the test locally against master and it passed. So, either your change is causing the failure, or some past commit (that has since been corrected) caused the failure and you need to merge in the latest master into your branch.

Documentation on how the test formatting is here: https://github.com/elastic/elasticsearch/blob/master/rest-api-spec/src/main/resources/rest-api-spec/test/README.asciidoc

From what I can see, it is a simple re-ordering issue.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/GetJobsStatsAction.java [hunk 1]
```diff
@@ -118,7 +118,12 @@ public boolean allowNoJobs() {
 
         @Override
         public boolean match(Task task) {
-            return OpenJobAction.JobTaskMatcher.match(task, jobId);
+            for(int i = 0; i < expandedJobsIds.size(); i++)
```
[the remark this answers is not in the document]
Comment (up201608320) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/GetJobsStatsAction.java]: Will commit the changes thx for the review!

**Comment by davidkyle** [comment]
The failing test in `job_groups.yml` is because the response of GET job stats is not ordered by job id. Compare the 6.x branch https://github.com/elastic/elasticsearch/blob/6.x/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportGetJobsStatsAction.java#L163

with master https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportGetJobsStatsAction.java#L151 where the call to `Collections.sort` is missing.

Actually this call is also missing from the master branch https://github.com/elastic/elasticsearch/blob/6.x/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportGetJobsStatsAction.java#L100

I assume this change has somehow affected the ordering causing the tests to fail which is good a is has revealed an issue. 

As for testing this change a unit test added to `GetJobStatsActionRequestTests` would be sufficient. There is already one test on the match function `testMatch_GivenAll_FailsForNonJobTasks` in that class. Note the function expects a Task parameter which implements`OpenJobAction.JobTaskMatcher`, use `TransportOpenJobAction.JobTask` for this

**Comment by davidkyle** [comment]
@up201608320 I raised #36841 to fix the ordering of the GET job stats response. Once that is merged please rebase your change on the master branch and the test failure will go away

**Comment by droberts195** [comment]
@up201608320 #36841 is now merged.  If you pull the latest master branch, then merge it into your PR branch then hopefully this PR will pass CI and can be merged.

**Comment by droberts195** [comment]
Jenkins test this please

**Comment by droberts195** [comment]
Jenkins run elasticsearch-ci/default-distro

**Comment by droberts195** [comment]
Jenkins run elasticsearch-ci/default-distro

**Comment by droberts195** [comment]
Jenkins test this please

**Comment by droberts195** [comment]
Jenkins test this please

**Review by davidkyle — APPROVED** [review]
LGTM

**Pull request merged**